### PR TITLE
Lock Route Table / Subnets

### PR DESCRIPTION
**Pull request by whiskeyjay**
With certain configurations, when a Route Table is being associated to a Subnet (create or update), a Route could be simultaneously added to the Route Table, causing a 429 error on the Subnet operation.

The fix is to lock the Route Table making sure when the Subnet is created / updated, it will be the only access to the Route Table. - The Route creation/update code is already doing the same thing.

### CODE REVIEW

File: builtin/providers/azurerm/resource_arm_subnet.go [hunk 1]
```diff
@@ -96,6 +96,15 @@ func resourceArmSubnetCreate(d *schema.ResourceData, meta interface{}) error {
 		properties.RouteTable = &network.RouteTable{
 			ID: &rtId,
 		}
+
+		routeTableName, err := parseRouteTableName(rtId)
```
Comment (StephenWeatherford) [inline builtin/providers/azurerm/resource_arm_subnet.go:108]: More idiomatic:
if routeTableName, err := parseRouteTableName(rtId); err != nil {
  return err
}

Comment (StephenWeatherford) [inline builtin/providers/azurerm/resource_arm_subnet.go:108]: More idiomatic:
if routeTableName, err := parseRouteTableName(rtId); err != nil {
  return err
}

File: builtin/providers/azurerm/resource_arm_subnet.go [hunk 2]
```diff
@@ -182,6 +191,18 @@ func resourceArmSubnetDelete(d *schema.ResourceData, meta interface{}) error {
 	name := id.Path["subnets"]
 	vnetName := id.Path["virtualNetworks"]
 
+	if v, ok := d.GetOk("route_table_id"); ok {
+		rtId := v.(string)
+		routeTableName, err := parseRouteTableName(rtId)
```
Comment (StephenWeatherford) [inline builtin/providers/azurerm/resource_arm_subnet.go]: ditto

File: builtin/providers/azurerm/resource_arm_subnet.go [hunk 3]
```diff
@@ -182,6 +191,18 @@ func resourceArmSubnetDelete(d *schema.ResourceData, meta interface{}) error {
 	name := id.Path["subnets"]
 	vnetName := id.Path["virtualNetworks"]
 
+	if v, ok := d.GetOk("route_table_id"); ok {
+		rtId := v.(string)
```
Comment (StephenWeatherford) [inline builtin/providers/azurerm/resource_arm_subnet.go]: Does the schema guarantee this is a string?  Or do we need to check?

File: builtin/providers/azurerm/resourceid.go [hunk 4]
```diff
@@ -95,3 +95,13 @@ func parseAzureResourceID(id string) (*ResourceID, error) {
 
 	return idObj, nil
 }
+
+func parseRouteTableName(routeTableId string) (string, error) {
+	id, err := parseAzureResourceID(routeTableId)
```
Comment (StephenWeatherford) [inline builtin/providers/azurerm/resourceid.go]: ditto

File: builtin/providers/azurerm/resourceid.go [hunk 5]
```diff
@@ -95,3 +95,13 @@ func parseAzureResourceID(id string) (*ResourceID, error) {
...
+
+	if err != nil {
+		return "", fmt.Errorf("[ERROR] Unable to parse Route Table ID '%s': %+v", routeTableId, err)
```
Comment (StephenWeatherford) [inline builtin/providers/azurerm/resourceid.go]: No [ERROR]  (that's only for logging)

File: builtin/providers/azurerm/resourceid.go [hunk 6]
```diff
@@ -95,3 +95,13 @@ func parseAzureResourceID(id string) (*ResourceID, error) {
...
+	}
```
Comment (StephenWeatherford) [inline builtin/providers/azurerm/resourceid.go]: Use %s for errors

File: builtin/providers/azurerm/resourceid.go [hunk 7]
```diff
@@ -95,3 +95,13 @@ func parseAzureResourceID(id string) (*ResourceID, error) {
...
+
+	return id.Path["routeTables"], nil
```
Comment (StephenWeatherford) [inline builtin/providers/azurerm/resourceid.go:114]: What if id.Path["routeTables"] == ""?

**Review by StephenWeatherford — CHANGES_REQUESTED** [review]
Added comments

File: builtin/providers/azurerm/resource_arm_subnet.go (hunk 1 above)
Comment (whiskeyjay) [reply builtin/providers/azurerm/resource_arm_subnet.go:108]: routeTableName is needed in Lock operation below, can't limit its scope to the if { } block.

**Pull request closed without merging**